<?xml version="1.0"?>
<launch>

  <!-- Spawn your team here. -->
  <arg name="x1_name" default="X1" />
  <arg name="x2_name" default="X2" />
  <arg name="x3_name" default="X3" />
  <arg name="x4_name" default="X4" />

  <!-- Spawn X1 -->
  <include file="$(find x1_gazebo)/launch/spawn_x1.launch">
    <arg name="robot_namespace" value="$(arg x1_name)"/>
    <arg name="x" value="125"/>
    <arg name="y" value="1"/>
    <arg name="z" value="-19"/>
    <arg name="yaw" value="3.14159"/>
  </include>

  <!-- Spawn X2 -->
  <group ns="$(arg x2_name)">
    <include file="$(find x2_description)/launch/description.launch">
      <arg name="robot_namespace" value="$(arg x2_name)" />
    </include>
    <node name="urdf_spawner" pkg="gazebo_ros" type="spawn_model"
          args="-urdf -model $(arg x2_name) -param robot_description -x 0 -y 1 -z 0.2" />
    <include file="$(find x2_control)/launch/control.launch">
      <arg name="robot_namespace" value="$(arg x2_name)" />
    </include>
  </group>

  <!-- X3 quadcopter -->
  <group ns="$(arg x3_name)">
    <include file="$(find x3_description)/launch/spawn_x3.launch">
      <arg name="mav_name" value="X3" />
      <arg name="x" value="-1"/>
      <arg name="y" value="-1"/>
      <arg name="z" value="0.1"/>
    </include>
    <!-- Spawn a position controller -->
    <node name="lee_position_controller_node" pkg="rotors_control" type="lee_position_controller_node">
      <rosparam command="load" file="$(find rotors_gazebo)/resource/lee_controller_iris.yaml" />
      <rosparam command="load" file="$(find rotors_gazebo)/resource/iris.yaml" />
      <remap from="odometry" to="odometry_sensor1/odometry" />
    </node>
    <!-- Give robot_descriptom parameter and joint states, publish TF frames -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
    <node name="robot_state_publisher_fixed" pkg="robot_state_publisher" type="robot_state_publisher">
        <remap from="joint_states" to="fixed/joint_states" />
    </node>
    <!-- Convert cmd_vel into pose command so as to provide the same interface -->
    <node name="cmdvel_transform" pkg="subt_example" type="cmdvel_converter.py" args="/X3/odometry_sensor1/pose /X3/cmd_vel /X3/command/pose"/>
  </group>

  <!-- X4 quadcopter -->
  <group ns="$(arg x4_name)">
    <include file="$(find x4_description)/launch/spawn_x4.launch">
      <arg name="mav_name" value="X4" />
      <arg name="x" value="-2"/>
      <arg name="y" value="-1"/>
      <arg name="z" value="0.1"/>
    </include>
    <!-- Spawn a position controller -->
    <node name="lee_position_controller_node" pkg="rotors_control" type="lee_position_controller_node">
      <rosparam command="load" file="$(find rotors_gazebo)/resource/lee_controller_neo11.yaml" />
      <rosparam command="load" file="$(find rotors_gazebo)/resource/neo11.yaml" />
      <remap from="odometry" to="odometry_sensor1/odometry" />
    </node>
    <!-- Give robot_descriptom parameter and joint states, publish TF frames -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" />
    <node name="robot_state_publisher_fixed" pkg="robot_state_publisher" type="robot_state_publisher">
        <remap from="joint_states" to="fixed/joint_states" />
    </node>
    <!-- Convert cmd_vel into pose command so as to provide the same interface -->
    <node name="cmdvel_transform" pkg="subt_example" type="cmdvel_converter.py" args="/X4/odometry_sensor1/pose /X4/cmd_vel /X4/command/pose"/>
  </group>

  <!-- Competitor's control nodes -->
  <node name="$(arg x1_name)_control" pkg="subt_example" type="subt_example_node" args="$(arg x1_name) $(arg x1_name)" />
  <node name="$(arg x2_name)_control" pkg="subt_example" type="subt_example_node" args="$(arg x2_name) $(arg x2_name)" />
  <node name="$(arg x3_name)_control" pkg="subt_example" type="subt_example_node" args="$(arg x3_name) $(arg x3_name)" />
  <node name="$(arg x4_name)_control" pkg="subt_example" type="subt_example_node" args="$(arg x4_name) $(arg x4_name)" />

  <!-- Teleop node -->
  <arg name="joy_dev" default="/dev/input/js0" />
  <include file="$(find subt_example)/launch/teleop.launch">
    <arg name="joy_dev" default="$(arg joy_dev)" />
    <arg name="robot_config" value="$(find subt_example)/config/robot_config.yaml"/>
  </include>

</launch>
